File: test/system/ls.mjs
// The store's directories and the listing's layout, plus mkdir -p.
// Part of the system suite; test/run.mjs runs the cases in order and
// doc/Testing.md has the rules they run by.

import {
    chdir, fail, hasRootfs, output, prompt, row, rows, screen, submit, words,
} from "./harness.mjs";

export function check() {
    let s = screen();
    // /bin and /etc are directories in the store, unpacked at boot from the
    // archive tools/pack.py wrote at the end of the build.
    if (hasRootfs) {
        s = submit("clear", 1183);
        s = submit("cat /etc/motd", 1184);
        if (!rows(s).some((line) => line.startsWith("braam —")))
            fail(`/etc/motd did not read back: ${JSON.stringify(rows(s))}`);
        s = submit("clear", 1185);
        // Five names and no directory among them, in columns down the grid
        // rather than across it. version is the boot unpack's stamp.
        s = submit("ls /etc", 1186);
        const etc = output(s).join(" ").split(/ +/).filter(Boolean).sort();
        if (etc.join("|") !== "anchor|help|motd|repositories|version")
            fail(`/etc did not list its files: ${JSON.stringify(output(s))}`);
        // The README is at the root, where somebody arriving will see it. Not
        // the whole row: what else is at the top level moves with boot.cpp.
        // The root is where a directory is still marked as one.
        s = submit("clear", 1186.1);
        s = submit("ls /", 1186.2);
        if (!words(s).includes("README"))
            fail(`/ did not list the README: ${JSON.stringify(output(s))}`);
        if (!words(s).includes("etc/"))
            fail(`/ did not mark etc as a directory: ${JSON.stringify(output(s))}`);
    }

    // The layout itself, on a fixture the suite owns: /bin re-breaks whenever a
    // program is added. Sizes are chosen so name order, size order and reversed
    // order all differ. After vmstat, which counts the syscalls these make.
    submit("mkdir /home/t", 1186.30);
    submit("mkdir /home/t/sub", 1186.31);
    submit("echo 1 > /home/t/aaa", 1186.32); // 2 bytes
    submit("echo 123456 > /home/t/bb", 1186.33); // 7 bytes
    // Restamps aaa, so -t and -S disagree — and so `touch` on a file that is
    // already there is a moved mtime rather than a no-op.
    submit("touch /home/t/aaa", 1186.34);
    // Hidden unless -a asks: every case below that does not ask must read as if
    // these two were not there, -R's walk included.
    submit("echo x > /home/t/.dot", 1186.35);
    submit("mkdir /home/t/.d", 1186.36);

    // aaa=3, bb=2, sub/=4, so the column is 4 + 2 wide and three of them fit.
    const listing = (line, now) => {
        submit("clear", now);
        return output(submit(line, now + 0.005)).join("|");
    };
    const listings = [
        ["ls /home/t", "aaa   bb    sub/"],
        ["ls -1 /home/t", "aaa|bb|sub/"],
        // A directory keeps no mtime in OPFS and reports a dash, not 1970.
        ["ls -l /home/t", "total 2|file 2 Jun 19 20:14 aaa|file 7 Jun 19 20:14 bb|" +
            "dir  0            - sub/"],
        ["ls -lh /home/t", "total 2|file 2B Jun 19 20:14 aaa|file 7B Jun 19 20:14 bb|" +
            "dir  0B            - sub/"],
        ["ls -S /home/t", "bb    aaa   sub/"],
        // Newest first, which the touch above made disagree with -S.
        ["ls -t /home/t", "aaa   bb    sub/"],
        ["ls -r /home/t", "sub/  bb    aaa"],
        ["ls -rS /home/t", "sub/  aaa   bb"],
        ["ls -rt /home/t", "sub/  bb    aaa"],
        // The last of -S and -t wins, as the last of -l, -1 and -C does.
        ["ls -St /home/t", "aaa   bb    sub/"],
        ["ls -tS /home/t", "bb    aaa   sub/"],
        // -a reveals the two, in byte order, so a dot name sorts first.
        ["ls -a /home/t", ".d/   .dot  aaa   bb    sub/"],
        ["ls -a1 /home/t", ".d/|.dot|aaa|bb|sub/"],
        ["ls -aR /home/t",
            "/home/t:|.d/   .dot  aaa   bb    sub/||/home/t/.d:||/home/t/sub:"],
        ["ls -d /home/t", "/home/t/"],
        // Bundled, and a named directory gets no `total`.
        ["ls -dl /home/t", "dir  0            - /home/t/"],
        ["ls -- /home/t", "aaa   bb    sub/"],
        ["ls -R /home/t", "/home/t:|aaa   bb    sub/||/home/t/sub:"],
        ["ls -lR /home/t",
            "/home/t:|total 2|file 2 Jun 19 20:14 aaa|file 7 Jun 19 20:14 bb|" +
            "dir  0            - sub/||/home/t/sub:|total 0"],
        // A file operand prints before a directory one, in a block of its own.
        ["ls /home/t/aaa /home/t", "/home/t/aaa||/home/t:|aaa   bb    sub/"],
        // A pipe is one name per line; -C forces columns into one, at eighty.
        ["ls /home/t | head -n 2", "aaa|bb"],
        // -c is bytes rather than lines, and cuts one mid-name.
        ["ls /home/t | head -c 5", "aaa|b"],
        ["ls -C /home/t | cat", "aaa   bb    sub/"],
    ];
    let at = 1186.4;
    for (const [line, want] of listings) {
        const got = listing(line, at);
        at += 0.02;
        if (got !== want)
            fail(`\`${line}\` printed ${JSON.stringify(got)}, expected ${JSON.stringify(want)}`);
    }

    // A name is as wide as its codepoints, not its bytes: the grid puts one
    // codepoint in one cell, so padding by bytes shifts every column after an
    // accented name left by one. abcdef is six cells and six bytes, naïve five
    // cells and six, which is what tells the two apart.
    submit("mkdir /home/u", 1186.75);
    submit("touch /home/u/abcdef /home/u/naïve /home/u/xy", 1186.76);
    submit("clear", 1186.77);
    const wide = output(submit("ls /home/u", 1186.78)).join("|");
    if (wide !== "abcdef  naïve   xy")
        fail(`a UTF-8 name misaligned the columns: ${JSON.stringify(wide)}`);
    submit("rm -r /home/u", 1186.79);

    s = submit("clear", 1186.8);
    s = submit("ls -z /home/t", 1186.81);
    if (!output(s).some((line) => line.startsWith("ls: illegal option -- z")))
        fail(`an unknown flag said nothing: ${JSON.stringify(output(s))}`);
    if (!rows(s).includes(prompt(2)))
        fail(`an unknown flag left ${row(s, s.cursor_y)}, expected ${prompt(2)}`);

    // -h is this program's own, so only the long spelling asks. A bare ls
    // lists, and asking is stdout and 0.
    s = submit("clear", 1186.82);
    s = submit("ls --help", 1186.83);
    if (!output(s).includes("    ls [-1CRSadhlrt] [<path>...]") || !rows(s).includes(prompt(0)))
        fail(`ls --help printed ${JSON.stringify(output(s))}, expected the usage block`);
    s = submit("clear", 1186.84);
    s = submit("ls -h /home/t", 1186.85);
    if (output(s).some((line) => line.startsWith("Usage:")))
        fail(`ls -h printed a usage block: ${JSON.stringify(output(s))}`);

    submit("rm -r /home/t", 1186.9);

    // mkdir -p: the walk over the components, which is make_dir_all in
    // src/proc/io.cpp. Sys::MkDir is one level, so what is checked here is that
    // an existing directory is tolerated and a file in the way is not.
    {
        let mt = 1187;
        const mkdir = (line) => {
            submit("clear", (mt += 0.01));
            const s = submit(line, (mt += 0.01));
            return [output(s).join("|"), rows(s)];
        };
        const ok = (line) => {
            const [said, r] = mkdir(line);
            if (said !== "")
                fail(`\`${line}\` printed ${JSON.stringify(said)}, expected nothing`);
            if (!r.includes(prompt(0)))
                fail(`\`${line}\` left a status other than 0`);
        };
        const bad = (line, want, status = 1) => {
            const [said, r] = mkdir(line);
            if (!said.includes(want))
                fail(`\`${line}\` printed ${JSON.stringify(said)}, expected ${want}`);
            if (!r.includes(prompt(status)))
                fail(`\`${line}\` left a status other than ${status}`);
        };

        // One level at a time still refuses a missing parent.
        bad("mkdir /home/p/a/b", "not found");
        ok("mkdir -p /home/p/a/b");
        // Idempotent: every component of it stands now.
        ok("mkdir -p /home/p/a/b");
        // Without -p the leaf must not exist, which -p has not changed.
        bad("mkdir /home/p", "already exists");
        // A file where the leaf goes is an error even with -p, which is the
        // stat make_dir_all does when the whole path was already there.
        submit("touch /home/p/f", (mt += 0.01));
        bad("mkdir -p /home/p/f", "already exists");
        bad("mkdir -p /home/p/f/g", "mkdir: /home/p/f/g: ");
        // Relative, so the walk's prefixes resolve against the cwd.
        submit("cd /home/p", (mt += 0.01));
        chdir("/home/p");
        ok("mkdir -p r/s/t");
        const [seen] = mkdir("ls -d r/s/t");
        if (seen !== "r/s/t/")
            fail(`a relative mkdir -p left ${JSON.stringify(seen)}`);
        submit("cd /home", (mt += 0.01));
        chdir("/home");
        bad("mkdir -z /home/p", "mkdir: illegal option -- z", 2);
        submit("rm -r /home/p", (mt += 0.01));
    }
}

File: src/cmd/ls.cpp
#include "fs/path.h"
#include "kernel/alloc.h"
#include "kernel/fmt.h"
#include "kernel/text.h"
#include "kernel/traits.h"
#include "proc/file.h"
#include "proc/opt.h"
#include "proc/time.h"
#include "proc/usage.h"

// A listing, laid out for whatever stdout turns out to be: columns on the
// terminal, one name per line into a pipe. The trailing slash on a directory is
// not optional dressing — it is the only thing distinguishing a directory from
// a file in the short form.

namespace {

constexpr Opts LS_OPTS{ "1CRSadhlrt", "" };

constexpr usize GAP         = 2;   // between columns
constexpr u32 WIDTH_DEFAULT = 80;  // -C with no terminal to measure
constexpr u64 BLOCK         = 512; // FS_BLOCK, which is what `total` counts
constexpr Str USAGE =
    "Usage:\n"
    "    ls [-1CRSadhlrt] [<path>...]\n"
    "Options:\n"
    "    -l    one line each: kind, size, time, name\n"
    "    -C    in columns; the default on a terminal\n"
    "    -1    one name per line\n"
    "    -R    descend into the directories listed\n"
    "    -a    names beginning with a dot as well\n"
    "    -d    the directory itself, not what is in it\n"
    "    -t    newest first\n"
    "    -S    largest first\n"
    "    -r    reverse whichever order is in force\n"
    "    -h    scale the sizes for a reader\n";

// "Mmm DD HH:MM" or "Mmm DD  YYYY", the two BSD forms, and the width of both.
constexpr usize STAMP_W  = 12;
constexpr i64 STAMP_NEAR = 15778476; // six months in seconds

// Everything that outlives an await. A coroutine frame past 512 bytes costs a
// whole 64 KiB span, so the entries, the recursion and the row being built live
// in one heap block instead.
struct Lister {
    bool detail  = false; // -l
    bool columns = false; // -C, or a terminal with no layout flag given
    bool human   = false; // -h
    bool recurse = false; // -R
    bool all     = false; // -a
    bool reverse = false; // -r
    char order   = 0;     // -S or -t, the last of the two given
    bool self    = false; // -d
    bool head    = false; // print `path:` before each directory

    u32 width = WIDTH_DEFAULT;

    // The wall clock -l renders against: which side of UTC to show a stamp in,
    // and what "recent" is measured from. Read once, and only for -l.
    i64 now_secs = 0;
    i32 tz_min   = 0;

    // The directory `ents` came from, for -l to name a link. Empty for the
    // operand block, whose names are whole paths already.
    String dir;

    Vec<DirEntry> ents; // the block being printed
    Vec<String> dirs;   // the directory operands, in the order given
    Vec<String> stack;  // -R: directories still to list, deepest last
    String row;         // one line, reused
    bool wrote = false; // has anything gone out, for the blank line before a header
};

// Cells, not bytes: the grid puts one codepoint in one cell, so Buf's put_left
// and put_right — which pad by size() — cannot be used on a name.
usize disp_width(Str s)
{
    usize n = 0;
    for (usize i = 0; i < s.size();) {
        char32_t ch = 0;
        usize k     = utf8_decode(s, i, ch);
        i += k ? k : 1;
        n++;
    }
    return n;
}

// The suffix a name carries: '/' for a directory, '@' for a symbolic link.
char name_mark(const DirEntry &e)
{
    if (e.kind == SYS_KIND_DIR)
        return '/';
    return e.kind == SYS_KIND_LINK ? '@' : 0;
}

usize entry_width(const DirEntry &e)
{
    return disp_width(e.name.str()) + (name_mark(e) ? 1 : 0);
}

bool put_name(String &out, const DirEntry &e)
{
    if (!out.append(e.name.str()))
        return false;
    char mark = name_mark(e);
    return !mark || out.push(mark);
}

bool pad(String &out, usize n)
{
    for (usize i = 0; i < n; i++)
        if (!out.push(' '))
            return false;
    return true;
}

// 10G, 121M, 512B — one decimal below ten, as df prints it.
void human(Buf<32> &b, u64 bytes)
{
    constexpr Str UNIT[] = { "B", "K", "M", "G", "T", "P" };
    usize u              = 0;
    u64 whole = bytes, rem = 0;
    while (whole >= 1024 && u + 1 < sizeof(UNIT) / sizeof(UNIT[0])) {
        rem = whole % 1024;
        whole /= 1024;
        u++;
    }
    b.put(whole);
    if (u && whole < 10) {
        u64 tenth = (rem * 10) / 1024;
        if (tenth)
            b.put('.').put(tenth);
    }
    b.put(UNIT[u]);
}

void put_size(Buf<32> &b, u64 bytes, bool human_sizes)
{
    if (human_sizes)
        human(b, bytes);
    else
        b.put(bytes);
}

void put2(Buf<64> &b, u32 v)
{
    b.put(char('0' + (v / 10) % 10)).put(char('0' + v % 10));
}

// One -l time column, right-aligned in STAMP_W. A file whose filesystem keeps
// no mtime gets a dash rather than 1970.
void put_stamp(Buf<64> &b, const Lister &st, u64 mtime)
{
    if (!mtime) {
        for (usize i = 1; i < STAMP_W; i++)
            b.put(' ');
        b.put('-');
        return;
    }

    i64 secs = i64(mtime / 1000) + st.tz_min * 60;
    Civil c  = civil(secs);
    b.put(TIME_MONTHS[c.month - 1]).put(' ');
    put2(b, c.day);
    b.put(' ');

    // The clock's own zone, so "recent" is measured the way it is displayed.
    i64 age = st.now_secs + st.tz_min * 60 - secs;
    if (age >= 0 && age < STAMP_NEAR) {
        put2(b, c.hour);
        b.put(':');
        put2(b, c.min);
    } else {
        b.put(' ').put(u32(c.year));
    }
}

// The default is name order, which vfs_list already delivered. -S and -t are
// stable, so that order stays the tiebreak; -r reverses whatever came out.
void sort_block(Lister &st)
{
    Vec<DirEntry> &v = st.ents;
    for (usize i = 1; st.order && i < v.size(); i++)
        for (usize k = i; k > 0; k--) {
            bool ahead = st.order == 'S' ? v[k].size > v[k - 1].size : v[k].mtime > v[k - 1].mtime;
            if (!ahead)
                break;
            swap(v[k], v[k - 1]);
        }
    if (st.reverse && v.size() > 1)
        for (usize i = 0, j = v.size() - 1; i < j; i++, j--)
            swap(v[i], v[j]);
}

// One shared column width, as BSD's is — not a width per column.
void plan(const Lister &st, usize n, usize &cols, usize &rows, usize &colw)
{
    usize w = 0;
    for (usize i = 0; i < n; i++)
        w = max(w, entry_width(st.ents[i]));

    colw = w + GAP;
    cols = (st.width + GAP) / colw; // the last column needs no trailing gap
    if (cols < 1)
        cols = 1;
    if (cols > n)
        cols = n;
    rows = (n + cols - 1) / cols;
    cols = (n + rows - 1) / rows; // shrink, so no column comes out empty
}

// What a failed write exits with: ^C stays 130, as it does everywhere else.
i32 write_bad(Error e)
{
    return e == Error::Cancelled ? 130 : 1;
}

Task<i32> emit_short(Lister &st)
{
    usize n = st.ents.size();
    if (!n)
        co_return 0;
    usize cols = 1, rows = n, colw = 0;
    if (st.columns)
        plan(st, n, cols, rows, colw);

    // Column-major: a sorted listing is read down a column.
    for (usize r = 0; r < rows; r++) {
        st.row.clear();
        for (usize c = 0; c < cols; c++) {
            usize i = c * rows + r;
            if (i >= n)
                break;
            if (!put_name(st.row, st.ents[i]))
                co_return 1;
            // The last cell of a row is not padded: trailing blanks would put
            // the cursor in the deferred-wrap column.
            if (i + rows < n && !pad(st.row, colw - entry_width(st.ents[i])))
                co_return 1;
        }
        if (!st.row.push('\n'))
            co_return 1;
        if (Result<void> w = co_await File::stdout().write(st.row.str()); w.is_err())
            co_return write_bad(w.error());
    }
    co_return 0;
}

Task<i32> emit_long(Lister &st, bool with_total)
{
    // The size column is as wide as this block needs, and `total` counts
    // 512-byte blocks. A directory reports no size through this filesystem, so
    // it contributes none.
    u64 total = 0;
    usize w   = 1;
    for (const DirEntry &e : st.ents) {
        total += (e.size + BLOCK - 1) / BLOCK;
        Buf<32> t;
        put_size(t, e.size, st.human);
        w = max(w, t.str().size());
    }

    if (with_total) {
        Buf<32> t;
        t.put("total ").put(total).put('\n');
        if (Result<void> x = co_await File::stdout().write(t.str()); x.is_err())
            co_return write_bad(x.error());
    }

    for (const DirEntry &e : st.ents) {
        Buf<32> num;
        put_size(num, e.size, st.human);

        // A size and a stamp are ASCII, so put_right's byte padding is cell
        // padding here.
        Buf<64> t;
        t.put(e.kind == SYS_KIND_DIR    ? Str("dir  ")
              : e.kind == SYS_KIND_LINK ? Str("link ")
                                        : Str("file "));
        t.put_right(num.str(), w).put(' ');
        put_stamp(t, st, e.mtime);
        t.put(' ');

        st.row.clear();
        if (!st.row.append(t.str()) || !put_name(st.row, e))
            co_return 1;

        // Read rather than resolved, so a dangling link still prints a target.
        if (e.kind == SYS_KIND_LINK) {
            String full;
            if (st.dir.empty() ? !full.assign(e.name.str())
                               : path_join(st.dir.str(), e.name.str(), full).is_err())
                co_return 1;
            if (Task<Result<String>> t2 = read_link(full.str())) {
                Result<String> got = co_await t2;
                if (got.is_err() && got.error() == Error::Cancelled)
                    co_return 130;
                if (got.is_ok() && (!st.row.append(" -> ") || !st.row.append(got.value().str())))
                    co_return 1;
            }
        }
        if (!st.row.push('\n'))
            co_return 1;
        if (Result<void> x = co_await File::stdout().write(st.row.str()); x.is_err())
            co_return write_bad(x.error());
    }
    co_return 0;
}

Task<i32> emit_block(Lister &st, bool with_total)
{
    if (st.ents.empty() && !with_total)
        co_return 0;
    st.wrote = true;

    Task<i32> t = st.detail ? emit_long(st, with_total) : emit_short(st);
    if (!t)
        co_return 1;
    co_return co_await t;
}

// A `path:` header, with a blank line before it once anything has been printed.
Task<i32> emit_head(Lister &st, Str path)
{
    st.row.clear();
    if (st.wrote && !st.row.push('\n'))
        co_return 1;
    if (!st.row.append(path) || !st.row.append(":\n"))
        co_return 1;
    if (Result<void> w = co_await File::stdout().write(st.row.str()); w.is_err())
        co_return write_bad(w.error());
    st.wrote = true;
    co_return 0;
}

// One directory: its entries, sorted, with -R's children pushed for later.
Task<i32> list_one(Lister &st, Str path)
{
    if (st.head) {
        Task<i32> h = emit_head(st, path);
        if (!h)
            co_return 1;
        if (i32 bad = co_await h)
            co_return bad;
    }

    Result<Vec<DirEntry>> r = Err(Error::NoMemory);
    if (Task<Result<Vec<DirEntry>>> t = list_dir(path))
        r = co_await t;
    if (r.is_err()) {
        if (r.error() == Error::Cancelled)
            co_return 130;
        if (Task<void> e = errln("ls", path, r.error()))
            co_await e;
        co_return 1;
    }

    st.ents = move(r.value());
    if (!st.dir.assign(path))
        co_return 1;

    // A leading dot has to be asked for. Dropped before the sort and before
    // -R's push, so a dot directory is neither printed nor walked.
    if (!st.all) {
        usize w = 0;
        for (usize i = 0; i < st.ents.size(); i++) {
            if (st.ents[i].name.size() && st.ents[i].name[0] == '.')
                continue;
            if (w != i)
                st.ents[w] = move(st.ents[i]);
            w++;
        }
        st.ents.resize(w);
    }
    sort_block(st);

    // Pushed in reverse of the order they will print, so they pop in it. A link
    // is not SYS_KIND_DIR, so -R cannot walk through one.
    if (st.recurse)
        for (usize i = st.ents.size(); i > 0; i--) {
            const DirEntry &e = st.ents[i - 1];
            if (e.kind != SYS_KIND_DIR)
                continue;
            String child;
            if (path_join(path, e.name.str(), child).is_err() || !st.stack.push(move(child))) {
                if (Task<void> x = errln("ls", path, Error::NoMemory))
                    co_await x;
                co_return 1;
            }
        }

    Task<i32> t = emit_block(st, st.detail);
    if (!t)
        co_return 1;
    co_return co_await t;
}

// Every operand is stat'd up front. Files print as one block, then each
// directory in the order it was given — which is what puts `ls f d` in the
// order BSD puts it in.
Task<i32> run(Lister &st, Args paths)
{
    i32 status  = 0;
    usize count = paths.size();

    for (usize k = 0; k == 0 || k < count; k++) {
        Str p = count ? paths[k] : Str(".");

        // -l reports an operand that is a link rather than its target; without
        // it a link to a directory is followed and listed.
        Result<FileInfo> s = Err(Error::NoMemory);
        if (Task<Result<FileInfo>> t = stat_of(p, !st.detail))
            s = co_await t;
        if (s.is_err()) {
            if (s.error() == Error::Cancelled)
                co_return 130;
            status = 1;
            if (Task<void> e = errln("ls", p, s.error()))
                co_await e;
            continue;
        }

        bool dir = s.value().kind == SYS_KIND_DIR && !st.self;
        String name;
        if (!name.assign(p))
            co_return 1;
        bool ok = dir ? st.dirs.push(move(name))
                      : st.ents.push(DirEntry{ move(name), s.value().kind, s.value().size,
                                               s.value().mtime });
        if (!ok)
            co_return 1;
    }

    st.head = count > 1 || st.recurse;

    if (!st.ents.empty()) {
        sort_block(st);
        Task<i32> t = emit_block(st, false);
        if (!t)
            co_return 1;
        if (i32 bad = co_await t) {
            if (bad == 130)
                co_return 130;
            status = 1;
        }
    }

    for (usize k = 0; k < st.dirs.size(); k++) {
        if (!st.stack.push(move(st.dirs[k])))
            co_return 1;
        // Depth first: -R pushed this one's children while it was listed.
        while (!st.stack.empty()) {
            String p = move(st.stack[st.stack.size() - 1]);
            st.stack.pop();
            Task<i32> t = list_one(st, p.str());
            if (!t)
                co_return 1;
            if (i32 bad = co_await t) {
                if (bad == 130)
                    co_return 130;
                status = 1;
            }
        }
    }
    co_return status;
}

} // namespace

Task<i32> proc_main(Args args)
{
    // -h is this program's own, so only the long spelling asks.
    if (args.size() == 2 && args[1] == "--help")
        co_return co_await usage_asked(USAGE);

    Lister *st = heap_new<Lister>();
    if (!st) {
        co_await write_all(SYS_STDERR, "ls: out of memory\n");
        co_return 1;
    }
    struct Free {
        ~Free() { heap_delete(p); }
        Lister *p;
    } free_lister{ st };

    // The last of -l, -1 and -C wins, as it does in BSD.
    char layout = 0;
    OptParse opts(args, LS_OPTS);
    Opt o;
    for (;;) {
        Result<bool> more = opts.next(o);
        if (more.is_err()) {
            Buf<64> b;
            b.put("ls: illegal option -- ").put(o.name).put('\n');
            co_await write_all(SYS_STDERR, b.str());
            co_return co_await usage_error(USAGE);
        }
        if (!more.value())
            break;
        switch (o.name) {
        case '1':
        case 'C':
        case 'l':
            layout = o.name;
            break;
        case 'R':
            st->recurse = true;
            break;
        case 'a':
            st->all = true;
            break;
        case 'S':
        case 't':
            st->order = o.name;
            break;
        case 'd':
            st->self = true;
            break;
        case 'h':
            st->human = true;
            break;
        case 'r':
            st->reverse = true;
            break;
        }
    }
    // -d lists a directory as itself, so there is nothing below it to walk.
    if (st->self)
        st->recurse = false;

    // The one question a program cannot answer for itself.
    Result<TtyInfo> tty = Err(Error::Unsupported);
    if (Task<Result<TtyInfo>> t = tty_of(SYS_STDOUT))
        tty = co_await t;
    if (tty.is_err() && tty.error() == Error::Cancelled)
        co_return 130;

    bool console = tty.is_ok() && tty.value().console;

    // The buffering ls would otherwise make File probe for a second time.
    File &out = File::stdout();
    out.set_buffering(console ? Buffering::Line : Buffering::Full);

    st->detail  = layout == 'l';
    st->columns = layout == 'C' || (!layout && console);
    if (console && tty.value().at.cols)
        st->width = tty.value().at.cols;

    // Only -l renders a stamp, and a clock that will not answer costs the
    // recent form rather than the listing.
    if (st->detail) {
        Result<Clock> clock = Err(Error::Unsupported);
        if (Task<Result<Clock>> t = clock_now())
            clock = co_await t;
        if (clock.is_err() && clock.error() == Error::Cancelled)
            co_return 130;
        if (clock.is_ok()) {
            st->now_secs = i64(clock.value().epoch_ms / 1000);
            st->tz_min   = clock.value().tz_min;
        }
    }

    Task<i32> t = run(*st, opts.rest());
    if (!t)
        co_return 1;
    i32 status = co_await t;

    if (Result<void> w = co_await out.flush(); w.is_err())
        co_return write_bad(w.error());
    co_return status;
}
